### PR TITLE
Features

### PR DESCRIPTION
**Pull request by UweBonnes**
- Handle Stlink V3
- Handle multiple connected Stlinks
- Try harder to handle devices much of the time in sleep
- Handle Stm32H7 programming
- Handle Stm32L4/G0  programming 
- Handle Stm32L0/L1 programming

### CODE REVIEW

File: lib/stm32devices.py [hunk 1]
```diff
@@ -141,6 +141,18 @@
                     {'type': 'STM32L083xZ', 'flash_size':  192, 'sram_size':  20, 'eeprom_size':  6, 'freq':  32},
                 ],
             },
+            {
+                'dev_id': 0x460,
+                'flash_size_reg': 0x1fff75e0,
+                'flash_driver': ,
+                'erase_sizes': (2048, ),
+                'devices': [
+                    {'type': 'STM32L070x8', 'flash_size':   64, 'sram_size':  32, 'eeprom_size':  0, 'freq':  364},
```
Comment (pavelrevak) [inline lib/stm32devices.py]: `L0` or `G0` ?

File: lib/stm32devices.py [hunk 2]
```diff
@@ -144,13 +144,13 @@
             {
                 'dev_id': 0x460,
                 'flash_size_reg': 0x1fff75e0,
-                'flash_driver': ,
-                'erase_sizes': (2048, ),
+                'flash_driver': 'STM32L4',
+                'erase_sizes': (2*1024, ),
```
Comment (pavelrevak) [inline lib/stm32devices.py]: spaces around `*`

File: lib/stm32l4.py [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+    FLASH_SR_WPRERR_BIT     = 1 <<  4
+    FLASH_SR_PGAERR_BIT     = 1 <<  5
+    FLASH_SR_SIZERR_BIT     = 1 <<  6
+    FLASH_SR_PGSERR_BIT     = 1 <<  7
+    FLASH_SR_FASTERR_BIT    = 1 <<  8
+    FLASH_SR_MISSERR_BIT    = 1 <<  9
+    FLASH_SR_BUSY_BIT       = 1 << 16
+    FLASH_SR_CFGBSY_BIT     = 1 << 18
+
+    FLASH_SR_ERROR_MASK \
```
Comment (pavelrevak) [inline lib/stm32l4.py]: do not use backslash `\`. put right side into brackets instead.

**Comment by UweBonnes** [comment]
Pavel Revak writes:
> 
> L0 or G0 ?

This is really STM32G0.

> +                'erase_sizes': (2*1024, ),
> 
> spaces around *

Done. Many mor entries miss that space!

> do not use backslash \. put right side into brackets instead.
Done.

Cheers.
-- 
Uwe Bonnes                bon@elektron.ikp.physik.tu-darmstadt.de

Institut fuer Kernphysik  Schlossgartenstrasse 9  64289 Darmstadt
--------- Tel. 06151 1623569 ------- Fax. 06151 1623305 ---------

File: lib/stm32h7.py [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+        self._dbg.bargraph_start('Writing FLASH', value_min=addr,
+                                 value_max=addr + len(data))
+        cr = Flash.FLASH_CR_PG | Flash.FLASH_CR_PSIZE32
+        if addr < 0x08100000:
+            flash.unlock(0)
+            self._stlink.set_debugreg32(Flash.FLASH_CR_REGS[0], cr)
+            cr =  self._stlink.get_debugreg32(Flash.FLASH_CR_REGS[0])
+            if not cr & Flash.FLASH_CR_PG:
+                raise lib.stlinkex.StlinkException(
+                    'Bank 0 FLASH_CR not ready for programming: %08x\n' % cr)
```
Comment (pavelrevak) [inline lib/stm32h7.py]: do not use `\n` in all exceptions

**Comment by pavelrevak** [comment]
if I trying to flash 2MB .bin file at once:
```
./pystlink flash:erase:verify:test.bin 
DEVICE: ST-Link/V2-1 V2J29M18
SUPPLY: 3.24V
CORE:   CortexM7
MCU:    STM32H743xI/STM32H753xI
FLASH:  2048KB
SRAM:   1024KB
Loaded 2097152 Bytes from test.bin file
*** Parameter error: name 'clear_sr' is not defined ***
```

**Comment by UweBonnes** [comment]
I removed all "\n" from exceptions . Now I use brackets instead of line continuation. clear_sr() is now self.clear_sr()

On first programming H7, I see  the error status register error 0x2010001 meaning 
Bit 25 SNECCERR1: Bank 1 single correction error flag
I do not see it with next flashing. Maybe some artefact from factory testing?

**Comment by pavelrevak** [comment]
Ok, I taking your PR, it has very important features, I will fix them later, thanks a lot!

**Comment by UweBonnes** [comment]
What still needs fixing?

**Pull request merged**